scripts/plotting/cloud_regime_analysis.py

```diff
@@ -0,0 +1,1369 @@
+                    tau_selection = {tau_var_name:slice(0,9999999999999)}
+                    # Making sure this works for pressure which is ordered largest to smallest and altitude which is ordered smallest to largest
+                    if ds_b[ht_var_name][0] > ds_b[ht_var_name][-1]: ht_selection = {ht_var_name:slice(9999999999999,0)}
+                    else: ht_selection = {ht_var_name:slice(0,9999999999999)}
+                    ds_b = ds_b.sel(tau_selection)
+                    ds_b = ds_b.sel(ht_selection)
+
+                    print(f'      -Preprocessing data') 
+
+                    # COSP ISCCP data has one extra tau bin than the satellite data, and misr has an extra height bin. This checks roughly if we are comparing against the 
+                    # satellite data, and if so removes the extra tau or ht bin. If a user passes home made CRs from CESM data, no data will be removed
+                    if data == 'ISCCP' and cl.shape[1] == 42:
+                            # A slightly hacky way to drop the smallest tau bin, but is robust incase tau is flipped in a future version
+                            ds_b = ds_b.sel(cosp_tau=slice(np.min(ds_b.cosp_tau)+1e-11,np.inf))
+                            print("      -Dropping smallest tau bin to be comparable with observational cloud regimes")
+                    if data == 'MISR' and cl.shape[1] == 105:
+                            # A slightly hacky way to drop the lowest height bin, but is robust incase height is flipped in a future version
+                            ds_b = ds_b.sel(cosp_htmisr=slice(np.min(ds_b.cosp_htmisr)+1e-11,np.inf))
+                            print("      -Dropping lowest height bin to be comparable with observational cloud regimes")
+
+                    # Selcting only the relevant data and stacking it to shape n_histograms, n_tau * n_pc
+                    dims = list(ds_b.dims)
+                    dims.remove(tau_var_name)
+                    dims.remove(ht_var_name)
+                    histograms_b = ds_b.stack(spacetime=(dims), tau_ht=(tau_var_name, ht_var_name))
+                    weights_b = np.cos(np.deg2rad(histograms_b.lat.values)) # weights_b array to use with emd-kmeans
+
+                    # Turning into a numpy array for clustering
+                    mat_b = histograms_b.values
+
+                    # Removing all histograms with 1 or more nans in them
+                    indicies = np.arange(len(mat_b))
+                    is_valid = ~np.isnan(mat_b.mean(axis=1))
+                    is_valid = is_valid.astype(np.int32)
+                    valid_indicies_b = indicies[is_valid==1]
+                    mat_b=mat_b[valid_indicies_b]
+                    weights_b=weights_b[valid_indicies_b]
+
+                    if np.min(mat_b < 0):
+                        raise Exception (f'Found negative value in ds_b.{var_name}, if this is a fill value for missing data, convert to nans and try again')
```

Comment:
Don't want to kill the ADF here:

```suggestion
                        print(f'Found negative value in ds_b.{var_name}, if this is a fill value for missing data, convert to nans and try again')
                        return
```